runtime/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -5216,12 +5204,50 @@ private void onSessionBegin(
 
                     sessionExpiry = mqttSessionBeginEx.expiry();
                     capabilities = mqttSessionBeginEx.capabilities();
-                    maximumQos = mqttSessionBeginEx.qosMax();
+                    subscribeMaximumQos = mqttSessionBeginEx.subscribeQosMax();
+                    if (mqttSessionBeginEx.packetIds() != null)
+                    {
+                        mqttSessionBeginEx.packetIds().forEachRemaining((IntConsumer) p -> unreleasedPacketIds.add(p));
+                    }
                 }
 
                 doSessionWindow(traceId, encodeSlotOffset, encodeBudgetMax);
             }
 
+            private void onSessionFlush(
+                FlushFW flush)
+            {
+                final long sequence = flush.sequence();
+                final long acknowledge = flush.acknowledge();
+                final long traceId = flush.traceId();
+                final long authorization = flush.authorization();
+                final OctetsFW extension = flush.extension();
+
+                assert acknowledge <= sequence;
+                assert sequence >= replySeq;
+
+                replySeq = sequence;
+
+                assert replyAck <= replySeq;
+
+                if (extension.sizeof() > 0)
+                {
+                    final MqttFlushExFW sessionFlushEx = extension.get(mqttFlushExRO::tryWrap);
+                    final int packetId = sessionFlushEx.session().packetId();
+
+                    unreleasedPacketIds.removeInt(packetId);
+                    switch (version)
+                    {
+                    case 4:
+                        doEncodePubcompV4(traceId, authorization, packetId);
+                        break;
+                    case 5:
+                        doEncodePubcompV5(traceId, authorization, packetId);
+                        break;
+                    }
```

Comment:
We will need to address all these switch statements for version 4 vs version 5 in the code each time an encoding decision is needed, even though the version is already decided just once up front.

We can use an inner class MqttVersion enum with methods like `doEncodePubcomp(...)` where the `v4` constant delegates to `doEncodePubcompV4` and the `v5` constant delegates to `doEncodePubcompV5`, while here we would just call `version.doEncodePubcomp(...)`. Similar approach for other `switch (version)` statements.

This feedback can be handled in a different PR though.